Code/Engine/Foundation/Platform/Win/DirectoryWatcher_Win.cpp

```diff
@@ -36,13 +39,246 @@ namespace
+    // Instead of building arrays of arrays, we create a bit field of all changes and then flatten the linked list at the first set bit. While iterating we remove everything we reached via the linked list so on the next call to get the first bit we will find another object that needs processing. As the operations are ordered, the first bit will always point to the very first operation of an object (nextOp can never point to a previous element).
+    ezBitfield<ezHybridArray<ezUInt32, 4>> pendingChanges;
+    pendingChanges.SetCount(ref_changes.GetCount(), true);
+
+    // Get start of first object.
+    ezHybridArray<Change*, 4> objectChanges;
+    auto it = pendingChanges.GetIterator();
+    while (it.IsValid())
+    {
+      // Flatten the changes for one object into a list for easier processing.
+      {
+        objectChanges.Clear();
+        ezUInt32 currentIndex = it.Value();
+        objectChanges.PushBack(&ref_changes[currentIndex]);
+        pendingChanges.ClearBit(currentIndex);
+        while (nextOp[currentIndex] != -1)
+        {
+          currentIndex = nextOp[currentIndex];
+          pendingChanges.ClearBit(currentIndex);
+          objectChanges.PushBack(&ref_changes[currentIndex]);
+        }
+      }
+
+      // Figure out what type the object is. There is no correct way of doing this, which is the reason why ReadDirectoryChangesExW exists. There are however some good heuristics we can use:
+      // 1. If the change is on an existing object, we should know it's type from the mirror. This is 100% correct.
+      // 2. If object still exists, we can query its stats on disk to determine its type. In rare cases, this can be wrong but you would need to create a case where a file is replaced with a folder of the same name or something.
+      // 3. If the object was created and deleted in the same enumeration, we cannot know what type it was, so we take a guess.
+      {
+        bool isFile = true;
+        bool typeFound = false;
+        for (Change* currentChange : objectChanges)
+        {
+          ezFileSystemMirrorType::Type type;
+          if (pMirror->GetType(currentChange->eventFilePath, type).Succeeded())
+          {
+            isFile = type == ezFileSystemMirrorType::Type::File;
+            typeFound = true;
+            break;
+          }
+          ezFileStats stats;
+          if (ezOSFile::GetFileStats(currentChange->eventFilePath, stats).Succeeded())
+          {
+            isFile = !stats.m_bIsDirectory;
+            typeFound = true;
+            break;
+          }
+        }
+
+        if (!typeFound)
+        {
+          // No stats and no entry in mirror: It's guessing time!
+          isFile = objectChanges[0]->eventFilePath.FindSubString(".") != nullptr;
+        }
+
+        // Apply type to all objects in the chain.
+        for (Change* currentChange : objectChanges)
+        {
+          currentChange->isFile = isFile;
+        }
```

Comment:
I don't understand the chain logic yet, but wouldn't you want to apply the change differently, depending on how confident you are ?

Ie when we know the result, apply it, if we don't know it and just guessed, store it, but wait for a better result? And don't override a result that was already certain?

Reply:
So a chain of events is something like:
1. "Create A -> Rename start A -> Rename end B -> Modify B". We know that all of these must refer to the same object but just looking at the first event "Create A" we can't figure out the type given that at the location 'A' no object exists anymore on disk and we have obviously not observed A in the mirror at that point as it is new.
If we can find a file at location B, then we know the entire chain must belong to a file. The only cases where this assumption is wrong are non-practical like creating a file and immediately deleting it and replacing it with a folder of the same name which will not appear in practise.
2. Another common change is something like "Modify C -> Modify C" or "Rename C -> Rename D". In these cases we have observed the object before so the very first op can be looked up in the mirror. Here we can always be certain that this is correct.
3. The only case that we have to guess is something like this: "Create A -> Delete A". Or cases where the monitor is overwhelmed and the kernel drops events.

As for waiting longer, I already flush all events that are in the kernel by running the code after the outer loop so no more events can be observed for these objects at the current time. So waiting any longer would probably not improve things are the state on disk will diverge even further from the observed events over time. And the question then also is when to stop and make a decision.

I'm not sure if there is a case in which waiting longer would result in a different decision. The file won't suddenly appear in the monitor and if we can't find it on disk anymore it will probably not come back anytime soon. There is of course always the possibility of race conditions between the ops on disk and the events in the monitor but the only case in which you would be able to trick the logic here would be if you rename a file multiple times in rapid succession so that the evaluation of the monitor results will not be able to observe the current location on disk as the file has already been renamed again, which is probably an unlikely case in practice.

The only way to improve this is probably making up heuristics based on the exact type of events in the chain but that would make the code a lot more complicated.
Before making any more detailed changes to the logic, I'd say we should wait for practical examples that we observe where things don't work as expected and then come up with better heuristics for these use cases.

Reply:
Okay makes sense.

Though I didn't mean "wait longer" like actually letting more time pass.

To me the code looks like it makes a decision based on different criteria, one of them is the fallback to look at the filename and check for dots. However, you have multiple events in your chain and one may give you a very certain answer, the other only the guessed answer, but to me it looks like you basically just take what you get and don't store how confident you are about it.

Ie if the first event gave you a perfect answer, and the second one is one where you have to guess, you shouldn't override the answer. The other way round, you SHOULD override the answer, because the first one wasn't as confident.

Probably your code is doing the right thing and I just don't see it, but that was what I meant with "wait for a better result".

Reply:
Ah I see what you mean, so this is happening implicitly as in the loop the mirror condition is checked first. I've changed the code now to make that clearer: Only the very first element of the chain needs to be looked up in the mirror as this is a list of ordered operations. If it's in the mirror, we are done. If not, we check all changes against the disk and if that also doesn't yield any results, we fall back to guessing.